drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,977 @@
+	}
+
+	frame->dlc = (msg_data.Header.Control & CANXL_TX_HEADER_DLC_MASK)
+						>> CANXL_TX_HEADER_DLC_SHIFT;
+
+	if (!!(msg_data.Header.Control & CANXL_TX_HEADER_FDF_MASK)) {
+		frame->flags |= CAN_FRAME_FDF;
+	}
+
+	if (!!(msg_data.Header.Control & CANXL_TX_HEADER_BRS_MASK)) {
+		frame->flags |= CAN_FRAME_BRS;
+	}
+
+	memcpy(frame->data, msg_data.data, can_dlc_to_bytes(frame->dlc));
+
+#ifdef CONFIG_CAN_RX_TIMESTAMP
+	frame->timestamp = msg_data.timeStampH << 32 | msg_data.timeStampL;
+#endif /* CAN_RX_TIMESTAMP */
+}
+
+static void can_nxp_s32_ctrl_callback(const struct device *dev,
+					Canexcel_Ip_EventType eventType, uint32 buffidx,
+					const Canexcel_Ip_StateType *canexcelState)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	struct can_frame frame = {0};
+	can_tx_callback_t tx_func;
+	can_rx_callback_t rx_func;
+	Canexcel_Ip_StatusType status;
+	Canexcel_Ip_DataInfoType tx_info;
+	int alloc;
+
+	if (eventType == CANEXCEL_EVENT_TX_COMPLETE) {
+		alloc = TX_MBIDX_TO_ALLOC_IDX(buffidx);
+		tx_func = data->tx_cbs[alloc].function;
+		tx_info = data->tx_cbs[alloc].tx_info;
+		LOG_DBG("%s: Sent Tx Mb %d", dev->name, buffidx);
+		if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+			tx_func(dev, 0, data->tx_cbs[alloc].arg);
+			k_sem_give(&data->tx_allocs_sem);
+		}
+	} else if (eventType == CANEXCEL_EVENT_RX_COMPLETE) {
+		alloc = RX_MBIDX_TO_ALLOC_IDX(buffidx);
+		rx_func = data->rx_cbs[alloc].function;
+		if (atomic_test_bit(data->rx_allocs, alloc)) {
+			nxp_s32_msg_data_to_zcan_frame(data->rx_msg[alloc], &frame);
+
+			LOG_DBG("%s: Received %d bytes Rx Mb %d, "
+				"Rx Id: 0x%x, "
+				"Id type: %s %s %s %s"
+				, dev->name, can_dlc_to_bytes(frame.dlc)
+				, buffidx, frame.id
+				, !!(frame.flags & CAN_FRAME_IDE) ?
+								"standard" : "extended"
+				, !!(frame.flags & CAN_FRAME_RTR) ? "RTR" : ""
+				, !!(frame.flags & CAN_FRAME_FDF) ? "FD frame" : ""
+				, !!(frame.flags & CAN_FRAME_BRS) ? "BRS" : "");
+
+			rx_func(dev, &frame, data->rx_cbs[alloc].arg);
+
+			status = Canexcel_Ip_ReceiveFD(config->instance, buffidx,
+							&data->rx_msg[alloc], FALSE);
+			if (status != CANEXCEL_STATUS_SUCCESS) {
+				LOG_ERR("Can't enable interrupt MB %d for next message!", buffidx);
+			}
+		}
+	}
+}
+
+static int can_nxp_s32_init(const struct device *dev)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	int err;
+
+	if (config->phy != NULL) {
+		if (!device_is_ready(config->phy)) {
+			LOG_ERR("CAN transceiver not ready");
+			return -ENODEV;
+		}
+	}
+
+	k_mutex_init(&data->rx_mutex);
+	k_sem_init(&data->tx_allocs_sem, CAN_NXP_S32_MAX_TX, CAN_NXP_S32_MAX_TX);
+
+	err = pinctrl_apply_state(config->pin_cfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Enable CANXL HW */
+	IP_MC_RGM->PRST_0[0].PRST_0 &=
+		~(MC_RGM_PRST_0_PERIPH_16_RST_MASK | MC_RGM_PRST_0_PERIPH_24_RST_MASK);
```

Comment:
Will this enable the CANXL controller and enable it to send ACKs? If so, this needs to be delayed until start() is called.

Reply:
If so, it needs to be enabled in listen-only mode so it does not disturb the bus with ACKs/error frames until explicitly enabled by calling `can_start()`.

Reply:
Please point me to where this is done?